src/tuner/nccl_ofi_model.c

```diff
@@ -74,167 +66,208 @@ static const float nccl_base_lat[NCCL_NUM_ALGORITHMS][NCCL_NUM_PROTOCOLS] = {
 
+	const double internodeBw = 1.0*NCCL_OFI_TUNER_INTERNODE_BW * chan/2 / dims->num_ranks;
+	const double intranodeBw = 1.0*NCCL_OFI_TUNER_INTRANODE_BW * dims->num_ranks;
 
-	const double intranode_latency = p2p_latency * num_steps_intranode;
-	const double intranode_bw = (NCCL_OFI_TUNER_INTRANODE_BW);
-	const double intranode_bytes = nBytes * (num_steps_intranode / num_steps);
-	const double intranode_cost = intranode_latency + (intranode_bytes / intranode_bw);
+	double effective_message_size = nBytes;
+	if (proto == NCCL_PROTO_LL)
+		effective_message_size *= 2;
+	else if (proto == NCCL_PROTO_LL)
+		effective_message_size += (nBytes / 15);
 
-	const double internode_latency = net_latency * num_steps_internode;
-	const double internode_bw = (NCCL_OFI_TUNER_INTERNODE_BW) * per_proto_bandwidth_overhead;
-	const double internode_bytes = nBytes * (num_steps_internode / num_steps);
-	const double internode_cost = internode_latency + (internode_bytes / internode_bw);
 
-	/* XXX: Should be scaled further here, not done for the sake of ensuring
-	 * this is cheaper than NVLSTree at dims where we did not explicitly opt
-	 * out of choosing ring above */
-	return ((base_latency * num_steps) + intranode_cost + internode_cost);
+	double totalCost = ((2.0 * effective_message_size / dims->num_ranks) * internodeSteps) / internodeBw;
+	totalCost += ((2.0*effective_message_size)*intranodeSteps)/intranodeBw;
+	totalCost += latencyCost * (get_ring_num_windows(2*effective_message_size, chan, dims->num_ranks, proto));
+	return totalCost;
 }
 
 double all_reduce_tree_cost(struct nccl_ofi_tuner_model_dims const* dims,
 			    const int proto,
 			    int pipe_ops,
+			    const size_t chan,
 			    const size_t nBytes)
 {
+
+
+
+	const double ranks_per_node = ceil(dims->num_ranks * 1.0 / dims->num_nodes);
+	if (1.0*chan/2.0 < ranks_per_node)
+		return INFINITY;
+
 	const double p2p_latency = nccl_nvlink_lat[NCCL_ALGO_TREE][proto];
 	const double base_latency = nccl_base_lat[NCCL_ALGO_TREE][proto];
-	const double num_steps_internode = log(dims->num_nodes) / log(2) - 1;
-	const double num_steps_intranode = log(dims->num_ranks) / log(dims->num_nodes) / log(2) - 1;
-	const double per_tree_msg_size = 2 * nBytes;
-	const double num_trees_per_rank = 2;
-	const double intranode_latency = p2p_latency * num_steps_intranode;
-	const double internode_latency = (NCCL_OFI_TUNER_NETWORK_LATENCY + base_latency) * num_steps_internode;
-	double latency = (intranode_latency + internode_latency);
+
+	const double num_trees_per_host = 2;
+	const double tree_height = floor(log(NCCL_OFI_IS_POWER_OF_TWO(dims->num_nodes) ? dims->num_nodes : dims->num_nodes - 1) / log(2)) + 1;
+
+	const double num_critical_steps_internode = /* there is a critical internode set of transfers */
+		2 * tree_height /* on the way up the tree, and again on the way back down. */
+		;
+
+	const double num_critical_steps_intranode = /* there is a critical-path intranode chain */
```

Comment:
in Tree, intranode steps are `(ranks_per_node - 1) * 2`. It seems `* tree_height` being added here to take latency overhead. However, it is not clear how the intranode (NV link) latency overhead is scaling with `tree_height`. If not, can we remove `tree_height` from here?

Reply:
same as other threads:

7b47266 origin/tuner-better-ring-nvlstree-boundary fixup! tuner: better ring rank/msize binning